parquet-column/src/main/java/org/apache/parquet/io/RecordReaderImplementation.java

```diff
@@ -424,15 +428,68 @@ public T read() {
     return record;
   }
 
+  private State getState(MessageType schema) {
+    for (State state : states) {
+        if (schema.containsField(state.fieldPath[0])) {
+          return state;
+        }
+    }
+    throw new IllegalArgumentException("Couldn't find state for schema " + schema);
+  }
+
+  protected void readVectors(ColumnVector[] vectors, MessageType[] columnSchemas, long current, long total) {
+      for (int i = 0 ; i < columnSchemas.length; i++) {
+        MessageType schema  = columnSchemas[i];
+        State state = getState(schema);
+        readVector(state.column, vectors[i], current, total);
+      }
+  }
+
+  protected void readVector(ObjectColumnVector<T> vector, long current, long total) {
```

Comment:
I think this API is pretty difficult to use. It looks like the purpose is to allow the caller to control this loop, but this passes information about the underlying reader back in from the outside.

I think a cleaner approach would be to delegate to the `ColumnReadStore` to return vectors in batches that align with record boundaries.